### PR TITLE
Backfill stripe metadata

### PR DESCRIPTION
**Pull request by arunagw**

### CODE REVIEW

File: app/models/payment_gateway_customer.rb [hunk 1]
```diff
@@ -43,6 +43,16 @@ def email
     def cards
       []
     end
+
+    def subscriptions
+      NoSubscription.new
+    end
```
Comment (gylaz) [inline app/models/payment_gateway_customer.rb]: Why is this necessary?

File: app/services/update_stripe_metadata.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
+class UpdateStripeMetadata
+  def self.run
+    Repo.where(private: true).find_each do |repo|
```
Comment (gylaz) [inline app/services/update_stripe_metadata.rb]: How about just `Repo.joins(:subscriptions)`?

File: app/models/payment_gateway_customer.rb (hunk 1 above)
Comment (arunagw) [inline app/models/payment_gateway_customer.rb]: This is required in one of test where customer do not have stripe_id. 

I will see if I can get rid of this

Comment (arunagw) [inline app/models/payment_gateway_customer.rb]: Should we remove this spec all together ? https://github.com/thoughtbot/hound/blob/aa-backfill-stripe-data/spec/services/update_stripe_metadata_spec.rb#L37-L51

**Comment by arunagw** [comment]
@gylaz @teoljungberg and I have also added tests for `PaymentGatewayCustomer` class as those were missing.

File: app/models/payment_gateway_customer.rb (hunk 1 above)
Comment (gylaz) [inline app/models/payment_gateway_customer.rb]: Ah, I understand it now. Let's keep the spec.

File: app/services/update_stripe_metadata.rb [hunk 3]
```diff
@@ -0,0 +1,20 @@
+class UpdateStripeMetadata
+  def self.run
+    Repo.joins(:subscription).find_each do |repo|
+      if subscription = repo.subscription
```
Comment (gylaz) [inline app/services/update_stripe_metadata.rb]: I don't think we need this `if` anymore. The query from line 3 guarantees a subscription, right?

File: spec/models/payment_gateway_customer_spec.rb [hunk 4]
```diff
@@ -14,4 +14,74 @@
       expect(update_request).to have_been_requested
     end
   end
+
+  describe "#email" do
+    it "will return customer email" do
+      user = build_stubbed :user, stripe_customer_id: stripe_customer_id
```
Comment (gylaz) [inline spec/models/payment_gateway_customer_spec.rb]: Parens around the arguments here for consistency?

File: spec/models/payment_gateway_customer_spec.rb [hunk 5]
```diff
@@ -14,4 +14,74 @@
...
+
+      customer = payment_gateway_customer.customer
+      expect(payment_gateway_customer.email).to eq customer.email
+    end
+  end
+
+  describe "#card_last4" do
+    context "when a customer has a card in stripe" do
+      it "returns last 4 from the customers card" do
+        user = build_stubbed :user, stripe_customer_id: stripe_customer_id
```
Comment (gylaz) [inline spec/models/payment_gateway_customer_spec.rb]: Parens.

File: spec/models/payment_gateway_customer_spec.rb [hunk 6]
```diff
@@ -14,4 +14,74 @@
       expect(update_request).to have_been_requested
     end
   end
+
+  describe "#email" do
+    it "will return customer email" do
```
Comment (gylaz) [inline spec/models/payment_gateway_customer_spec.rb]: We've been following the convention of `it "returns ...`.

File: spec/models/payment_gateway_customer_spec.rb [hunk 7]
```diff
@@ -14,4 +14,74 @@
...
+      user = build_stubbed :user, stripe_customer_id: stripe_customer_id
+      stub_customer_find_request
+
```
Comment (gylaz) [inline spec/models/payment_gateway_customer_spec.rb]: This newline is unnecessary.

File: spec/models/payment_gateway_customer_spec.rb [hunk 8]
```diff
@@ -14,4 +14,74 @@
...
+    end
+  end
+
+  describe "#card_last4" do
+    context "when a customer has a card in stripe" do
+      it "returns last 4 from the customers card" do
+        user = build_stubbed :user, stripe_customer_id: stripe_customer_id
+        stub_customer_find_request
+
+        payment_gateway_customer = PaymentGatewayCustomer.new(user)
```
Comment (gylaz) [inline spec/models/payment_gateway_customer_spec.rb]: This should be part of setup. Since we're not testing the constructor, separating this with newlines isn't necessary.

File: spec/models/payment_gateway_customer_spec.rb [hunk 9]
```diff
@@ -14,4 +14,74 @@
...
+
+        expect(payment_gateway_customer.card_last4).to eq "4242"
+      end
+    end
+
+    context "when a customer does not have a card in stripe" do
+      it "returns a null object" do
+        user = build_stubbed :user, stripe_customer_id: nil
+
+        payment_gateway_customer = PaymentGatewayCustomer.new(user)
```
Comment (gylaz) [inline spec/models/payment_gateway_customer_spec.rb]: Same here.

File: spec/models/payment_gateway_customer_spec.rb [hunk 10]
```diff
@@ -14,4 +14,74 @@
...
+
+        expect(payment_gateway_customer.card_last4).to eq ""
+      end
+    end
+  end
+
+  describe "#customer" do
+    context "when stripe_customer_id is present" do
+      it "retrieve customer data" do
+        user = build_stubbed :user, stripe_customer_id: stripe_customer_id
```
Comment (gylaz) [inline spec/models/payment_gateway_customer_spec.rb]: Parens.

File: spec/models/payment_gateway_customer_spec.rb [hunk 11]
```diff
@@ -14,4 +14,74 @@
       expect(update_request).to have_been_requested
     end
   end
+
+  describe "#email" do
+    it "will return customer email" do
+      user = build_stubbed :user, stripe_customer_id: stripe_customer_id
+      stub_customer_find_request
```
Comment (gylaz) [inline spec/models/payment_gateway_customer_spec.rb]: Since this is just testing delegations, what do you think about just stubbing `Stripe::Customer` here?

File: spec/models/payment_gateway_customer_spec.rb [hunk 12]
```diff
@@ -14,4 +14,74 @@
...
+      end
+    end
+
+    it "is memoized" do
+      user = build_stubbed :user, stripe_customer_id: nil
+
+      payment_gateway_customer = PaymentGatewayCustomer.new(user)
+      customer = payment_gateway_customer.customer
+
+      expect(customer).to be customer
```
Comment (gylaz) [inline spec/models/payment_gateway_customer_spec.rb]: I don't understand this test.

File: spec/services/update_stripe_metadata_spec.rb [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+    context "updates successfully" do
+      it "updates stripe metadata with repo id" do
+        user = create(:user, stripe_customer_id: stripe_customer_id)
+        repo = create(:repo, private: true)
+        subscription = create(
+          :subscription,
+          stripe_subscription_id: stripe_subscription_id,
+          user: user,
+          repo: repo
+        )
```
Comment (gylaz) [inline spec/services/update_stripe_metadata_spec.rb]: To cut down on some of the noise in this spec consider:
```rb
subscription = create(:subscription)
```
Then pass the subscription object to `stub_subscription_meta_data_update_request`.

File: spec/services/update_stripe_metadata_spec.rb [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+        UpdateStripeMetadata.run
+
+        expect(stripe_update_request).to have_been_requested
+      end
+    end
+
+    context "does not update" do
+      it "will not update if repo does not have subscription" do
+        create(:repo)
+        stripe_customer_find_request = stub_customer_find_request
```
Comment (gylaz) [inline spec/services/update_stripe_metadata_spec.rb]: I'm not sure this test is very valid. `stub_customer_find_request` could still fire with a different `stripe_customer_id` (other than the global).

File: spec/services/update_stripe_metadata_spec.rb [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+        )
+
+        UpdateStripeMetadata.run
+
+        expect(stripe_update_request).to have_been_requested
+      end
+    end
+
+    context "does not update" do
+      it "will not update if repo does not have subscription" do
```
Comment (gylaz) [inline spec/services/update_stripe_metadata_spec.rb]: `it "does not update ...`

File: spec/services/update_stripe_metadata_spec.rb [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+          repo.id
+        )
+
+        UpdateStripeMetadata.run
+
+        expect(stripe_update_request).to have_been_requested
+      end
+    end
+
+    context "does not update" do
```
Comment (gylaz) [inline spec/services/update_stripe_metadata_spec.rb]: `context` should provide the condition. In this case this should be:
```rb
context "when repo does not have a subscription" do
  it "does not make a request to Stripe" do
```

File: spec/services/update_stripe_metadata_spec.rb [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+      it "will not update if repo does not have subscription" do
+        create(:repo)
+        stripe_customer_find_request = stub_customer_find_request
+
+        UpdateStripeMetadata.run
+
+        expect(stripe_customer_find_request).not_to have_been_requested
+      end
+
+      it "will not update stripe when stripe_customer_id is missing" do
```
Comment (gylaz) [inline spec/services/update_stripe_metadata_spec.rb]:
```rb
context "when stripe_customer_id is missing" do
  it "does not make a request to Stripe" do
````

File: spec/services/update_stripe_metadata_spec.rb [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+        repo = create(:repo, private: true)
+        create(
+          :subscription,
+          user: user,
+          repo: repo
+        )
+
+        UpdateStripeMetadata.run
+
+        stripe_customer_find_request = stub_customer_find_request
```
Comment (gylaz) [inline spec/services/update_stripe_metadata_spec.rb]: The request still could have fired, but with different params. It might be best to verify that no requests to stripe were made:
```rb
stub_request(:any,%r{^https://api\.stripe\.com.+})
```

File: spec/support/helpers/stripe_api_helper.rb [hunk 19]
```diff
@@ -89,10 +91,24 @@ def stub_subscription_delete_request
     )
   end
 
+  def stub_subscription_meta_data_update_request(repo_id)
+    stub_request(
+      :post,
+      "#{stripe_base_url}/#{stripe_customer_id}/"\
+      "subscriptions/#{stripe_subscription_id}"
```
Comment (gylaz) [inline spec/support/helpers/stripe_api_helper.rb]: Is this missing an indentation?

File: spec/models/payment_gateway_customer_spec.rb (hunk 12 above)
Comment (teoljungberg) [inline spec/models/payment_gateway_customer_spec.rb]: Fixed and pushed, the test is supposed to check that the two customer's object_id are the same(hence they are memoized). I blame that it was on a friday afternoon

**Comment by arunagw** [comment]
This PR is ready for final review

File: spec/models/payment_gateway_customer_spec.rb [hunk 20]
```diff
@@ -14,4 +14,73 @@
       expect(update_request).to have_been_requested
     end
   end
+
+  describe "#email" do
+    it "returns customer email" do
+      user = build_stubbed(:user, stripe_customer_id: stripe_customer_id)
+      stub_customer_find_request
+
+      payment_gateway_customer = PaymentGatewayCustomer.new(user)
```
Comment (gylaz) [inline spec/models/payment_gateway_customer_spec.rb]: Why is this separated by newlines? This doesn't seem like the execution. Typically, the execution phase is the call to the method mentioned in the `describe`. It seems that this test has a combined execution and verification phases. I think it isn't helpful to mark this line as execution.

File: spec/models/payment_gateway_customer_spec.rb [hunk 21]
```diff
@@ -14,4 +14,73 @@
...
+    end
+  end
+
+  describe "#card_last4" do
+    context "when a customer has a card in stripe" do
+      it "returns last 4 from the customers card" do
+        user = build_stubbed(:user, stripe_customer_id: stripe_customer_id)
+        stub_customer_find_request
+
+        payment_gateway_customer = PaymentGatewayCustomer.new(user)
```
Comment (gylaz) [inline spec/models/payment_gateway_customer_spec.rb]: Same here.

File: spec/models/payment_gateway_customer_spec.rb [hunk 22]
```diff
@@ -14,4 +14,73 @@
...
+    end
+  end
+
+  describe "#customer" do
+    context "when stripe_customer_id is present" do
+      it "retrieve customer data" do
+        user = build_stubbed(:user, stripe_customer_id: stripe_customer_id)
+        stub_customer_find_request
+
+        payment_gateway_customer = PaymentGatewayCustomer.new(user)
```
Comment (gylaz) [inline spec/models/payment_gateway_customer_spec.rb]: And here.

File: spec/models/payment_gateway_customer_spec.rb [hunk 23]
```diff
@@ -14,4 +14,73 @@
...
+
+        expect(payment_gateway_customer.customer.id).to eq stripe_customer_id
+      end
+    end
+
+    context "when stripe_customer_id is not present" do
+      it "return null object" do
+        user = build_stubbed(:user, stripe_customer_id: nil)
+
```
Comment (gylaz) [inline spec/models/payment_gateway_customer_spec.rb]: Unnecessary newline.

File: spec/models/payment_gateway_customer_spec.rb [hunk 24]
```diff
@@ -14,4 +14,73 @@
...
+        expect(customer.email).to eq ""
+        expect(customer.cards).to eq []
+        expect(customer.subscriptions.retrieve).to eq nil
+      end
+    end
+
+    it "is memoized" do
+      user = build_stubbed(:user, stripe_customer_id: nil)
+
```
Comment (gylaz) [inline spec/models/payment_gateway_customer_spec.rb]: Unnecessary newline.

File: spec/services/update_stripe_metadata_spec.rb [hunk 25]
```diff
@@ -0,0 +1,51 @@
...
+          create(
+            :subscription,
+            user: user,
+            repo: repo
+          )
+
+          UpdateStripeMetadata.run
+
+          stripe_customer_find_request = stub_customer_find_request
+
```
Comment (gylaz) [inline spec/services/update_stripe_metadata_spec.rb]: Unnecessary newline.

**Comment by gylaz** [comment]
Left a few comments, then good to merge.
